dockstore-webservice/src/main/java/io/dockstore/webservice/languages/LanguageHandlerInterface.java

```diff
@@ -863,19 +850,37 @@ default String calculateDockerImageDigest(String manifestBody) {
      * Get an anonymous token with pull access to make Docker Registry HTTP API V2 calls.
      * Source for token request specs: https://docs.docker.com/registry/spec/auth/token/#requesting-a-token
      *
-     * @param registryDockerPath
+     * @param registry
      * @param repo
      * @return token
-     * @throws URISyntaxException
-     * @throws IOException
-     * @throws InterruptedException
      */
-    default String getDockerToken(String registryDockerPath, String repo) throws URISyntaxException, IOException, InterruptedException {
-        String getTokenURL = String.format("https://%s/token?scope=repository:%s:pull&service=%s", registryDockerPath, repo, registryDockerPath);
-        HttpRequest request = HttpRequest.newBuilder().uri(new URI(getTokenURL)).GET().build();
-        HttpResponse<String> tokenResponse = HttpClient.newBuilder().proxy(ProxySelector.getDefault()).build().send(request, HttpResponse.BodyHandlers.ofString());
+    default Optional<String> getDockerToken(Registry registry, String repo) {
+        String getTokenURL;
+        if (registry == Registry.AMAZON_ECR) {
+            getTokenURL = String.format("https://%s/token/?scope=repository:%s:pull&service=%s", registry.getDockerPath(), repo, registry.getDockerPath());
+        } else {
+            getTokenURL = String.format("https://%s/token?scope=repository:%s:pull&service=%s", registry.getDockerPath(), repo,
+                    registry.getDockerPath());
+        }
+
+        HttpResponse<String> tokenResponse;
+        try {
+            HttpRequest request = HttpRequest.newBuilder().uri(new URI(getTokenURL)).GET().build();
+            tokenResponse = HttpClient.newBuilder().proxy(ProxySelector.getDefault()).build().send(request, HttpResponse.BodyHandlers.ofString());
+        } catch (URISyntaxException | IOException | InterruptedException ex) {
+            LOG.error("Could not send token request GET {}", getTokenURL, ex);
+            return Optional.empty();
+        }
+
+        if (tokenResponse.statusCode() != HttpStatus.SC_OK) {
+            Map<String, List<Map<String, String>>> errorMap = GSON.fromJson(tokenResponse.body(), Map.class);
```

Comment:
Can you always rely on it returning a JSON response for all error conditions? If not, I think this line will throw an exception.

It may be so rare that you don't need to handle it; your call.

Reply:
The [Docker Registry API specs](https://docs.docker.com/registry/spec/api/#errors) say that actionable failure conditions are reported as part of 4xx responses in a JSON response body, so I think it's pretty safe to assume that it'll return a JSON response

Reply:
In theory I would think you could get 5xx responses as well, which may not be actionable, and hence may not have have JSON responses.

I tend to think in edge cases, and this one is probably way out there, so what you have should be fine.